posts/2022-01-03-kryo-pitfalls.adoc

```diff
@@ -0,0 +1,289 @@
+input.close()
+----
+
+This is perfectly valid, even if `Message` and `Person` are completely different, they have different field
+names. The only similarity is that both classes have three fields with the same types, just with a different
+order.
+The reason why this works has to do with how `FieldSerializer` (the default serializer for objects) is
+implemented:
+
+____
+FieldSerializer is efficient by writing only the field data, without any schema information, using the
+Java class files as the schema. It does not support adding, removing, or changing the type of fields
+without invalidating previously serialized bytes. Renaming fields is allowed only if it doesn't change
+the alphabetical order of the fields.
+____
+
+Since both classes' fields, once sorted by name, define the same order:
+`age, id, name` `==` `size, timestamp, wholePayload`,  Kryo deserializes a class' data into another class.
+We can solve this issue if we replace both write and read methods with ones that include the classes inside
+the payload:
+
+[source,scala]
+----
+case class Message(wholePayload: String, size: Int, ts: Long)
+case class Person(id: Long, name: String, age: Int)
+val alice = Person(1L, "Alice", 42)
+
+val kryo = new Kryo()
+kryo.register(classOf[Person])
+kryo.register(classOf[Message])
+
+val output = new Output(new FileOutputStream("file.bin"))
+kryo.writeClassAndObject(output, alice)
+output.close()
+
+val input = new Input(new FileInputStream("file.bin"))
+val personFromFile = kryo.readClassAndObject(input)
+input.close()
+----
+
+Notice that we no longer have to specify the `class` while reading. Thus we're effectively trading schema
+enforcement from _read_ time to _write_ time. Which coming full circle, if we attempt to do the same thing
+but without registering both classes, we get different outputs:
+
+[source,shell]
+----
+❯ xxd class-and-object-with-registration.bin
+00000000: 0c01 5402 0141 6c69 63e5                 ..T..Alic.
+
+❯ xxd class-and-object-without-registration.bin
+00000000: 0100 6578 616d 706c 652e 4261 7369 634b  ..example.BasicK
+00000010: 7279 6f55 7361 6765 5370 6563 2450 6572  ryoUsageSpec$Per
+00000020: 736f ee01 5402 0141 6c69 63e5            so..T..Alic.
+----
+
+The first serialization with registration includes the class identifier: `0c`. The rest of the payload is the
+same as the ones produced by the previous example. The second serialization uses the class' fully qualified
+name, which not only uses more space, but it's also less performant. You can run `RegistrationBenchmark` to
+compare the performance of both types, according to my numbers both speed and space increase between _10%_
+and _20%_.
+
+.Serializing 50k Messages
+|=====================================================================
+|                         |With Registration |Without Registration
+|With Class in Payload    | 657.663051 ms    | 767.096071 ms
+|Without Class in Payload | 600.054279 ms    | 763.372908 ms
```

Comment:
I would round these to at most 3 or 4 digits to reduce the noise:

```suggestion
|With Class in Payload    | 657.7 ms    | 767.1 ms
|Without Class in Payload | 600.1 ms    | 763.4 ms
```

Reply:
Good idea